Zend/zend_inheritance.c

```diff
@@ -1901,7 +1894,7 @@ static void zend_add_trait_method(zend_class_entry *ce, zend_string *name, zend_
 			 */
 			do_inheritance_check_on_method(
 				existing_fn, fixup_trait_scope(existing_fn, ce), fn, fixup_trait_scope(fn, ce),
-				ce, NULL, /* check_visibility */ 0);
+				ce, NULL, ZEND_INHERITANCE_CHECK_PROTO);
```

Comment:
I think this should also check visibility. This seems to be an existing bug.

https://3v4l.org/JjCIC

Reply:
I'm not sure. Private method shouldn't be visible in the child.

Anyway, my patch is mainly a clean-up.
I remember how we spent few hours reconstructing the logic of all these checks and I wouldn't like to repeat this work once again.

Reply:
I think, the related fixes and extension should be easier now.

Reply:
You are right!

I'm not sure why we disabled check for visibility before and if enabling it would cause different problems.
I'll check this, but anyway, I would prefer to make this change separately.

Reply:
I think it needs to account for private and private, which should succeed.

Reply:
Adding ``ZEND_INHERITANCE_CHECK_VISIBILITY`` breaks ``Zend/tests/traits/abstract_method_4.phpt``.
So it's better to add a separate check especially for trait.

Reply:
Oh, huh. So it looks like this is intentional. Let's keep that for 8.2 for sure. Adjusting this for master can be considered separately.

Reply:
I created an issue for this here. https://github.com/php/php-src/issues/14156 Let's leave it here as is.